### PR TITLE
Feature/generic repository

### PR DESCRIPTION
**Pull request by kenshi-takahashi**
Add generic repository and example for other repositories
Refactor: 
- Add IGenericRepository interface and implemented it in GenericRepository
- Add AsNoTracking() to GetAllAsync method
- Remove checks from GenericRepository

### CODE REVIEW

File: OnlineShop.DAL/Repositories/CategoryRepository.cs
Comment (kenshi-takahashi) [inline OnlineShop.DAL/Repositories/CategoryRepository.cs:1]: For example

File: OnlineShop.DAL/Interfaces/ICategoryRepository.cs
Comment (kenshi-takahashi) [inline OnlineShop.DAL/Interfaces/ICategoryRepository.cs:1]: For example

File: OnlineShop.DAL/Repositories/GenericRepository.cs [hunk 1]
```diff
@@ -0,0 +1,44 @@
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.EntityFrameworkCore;
+
+public class GenericRepository<T> where T : class
```
Comment (Elizaveta564) [inline OnlineShop.DAL/Repositories/GenericRepository.cs]: need to create IGenericRepository interface

File: OnlineShop.DAL/Repositories/GenericRepository.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+    public async Task<T> GetByIdAsync(int id)
+    {
+        return await _context.Set<T>().FindAsync(id);
+    }
+
+    public async Task<IEnumerable<T>> GetAllAsync()
+    {
+        return await _context.Set<T>().ToListAsync();
+    }
```
Comment (Elizaveta564) [inline OnlineShop.DAL/Repositories/GenericRepository.cs]: use AsNoTracking() for GET methods

File: OnlineShop.DAL/Repositories/GenericRepository.cs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+    public async Task DeleteAsync(int id)
+    {
+        var entity = await _context.Set<T>().FindAsync(id);
+        if (entity != null)
+        {
+            _context.Set<T>().Remove(entity);
+            await _context.SaveChangesAsync();
+        }
```
Comment (Elizaveta564) [inline OnlineShop.DAL/Repositories/GenericRepository.cs]: all checks should be in services and not in repositories

**Review by Elizaveta564 — CHANGES_REQUESTED** [review]

File: OnlineShop.DAL/Repositories/GenericRepository.cs [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+{
+    protected readonly OnlineShopDbContext _context;
+    public GenericRepository(OnlineShopDbContext context)
+    {
+        _context = context;
+    }
+
+    public async Task<T> GetByIdAsync(int id)
+    {
+        return await _context.Set<T>().FindAsync(id);
```
Comment (Elizaveta564) [inline OnlineShop.DAL/Repositories/GenericRepository.cs:13]: use AsNoTracking()

**Review by Elizaveta564 — APPROVED** [review]

File: OnlineShop.DAL/Repositories/GenericRepository.cs (hunk 4 above)
Comment (kenshi-takahashi) [reply OnlineShop.DAL/Repositories/GenericRepository.cs:13]: it is impossible to use AsNoTracking() here
image

**Pull request merged**